File: solutions/clojure/roman-numerals/1/src/roman_numerals.clj
(ns roman-numerals)

(require '[clojure.string :as str])

(defn numerals
  "Convert a number to its roman numeral(s)"
  [n]
  (-> (loop [n n
             result []]
        (cond
          (>= n 1000) (recur (- n 1000) (conj result "M"))
          (>= n  500) (recur (- n  500) (conj result "D"))
          (>= n  100) (recur (- n  100) (conj result "C"))
          (>= n   50) (recur (- n   50) (conj result "L"))
          (>= n   10) (recur (- n   10) (conj result "X"))
          (>= n    5) (recur (- n    5) (conj result "V"))
          (>= n    1) (recur (- n    1) (conj result "I"))
          :else (apply str result)))
      (str/replace "DCCCC" "CM")
      (str/replace  "CCCC" "CD")
      (str/replace "LXXXX" "XC")
      (str/replace  "XXXX" "XL")
      (str/replace "VIIII" "IX")
      (str/replace  "IIII" "IV")))

File: solutions/go/roman-numerals/1/roman_numerals.go
package romannumerals

import (
	"errors"
	"strings"
)

func ToRomanNumeral(input int) (string, error) {
	var result string

	if input < 1 || input > 3999 {
		return "", errors.New("Number is out of range")
	}
	for input > 0 {
		switch {
		case input >= 1000:
			result += "M"
			input -= 1000
		case input >= 500:
			result += "D"
			input -= 500
		case input >= 100:
			result += "C"
			input -= 100
		case input >= 50:
			result += "L"
			input -= 50
		case input >= 10:
			result += "X"
			input -= 10
		case input >= 5:
			result += "V"
			input -= 5
		case input >= 1:
			result += "I"
			input -= 1
		}
	}
	result = strings.Replace(result, "DCCCC", "CM", 1)
	result = strings.Replace(result, "CCCC", "CD", 1)
	result = strings.Replace(result, "LXXXX", "XC", 1)
	result = strings.Replace(result, "XXXX", "XL", 1)
	result = strings.Replace(result, "VIIII", "IX", 1)
	result = strings.Replace(result, "IIII", "IV", 1)
	return result, nil
}
